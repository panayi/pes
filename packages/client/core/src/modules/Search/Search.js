import React from 'react';
import * as R from 'ramda';
import { isNilOrEmpty } from 'ramda-adjunct';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { connectData } from '../../lib/connectData';
import { models } from '../../store/firebase/data';
import {
  actions as searchActions,
  selectors as searchSelectors,
} from '../../store/search';
import { selectors as pageSelectors } from '../../store/search/page';
import {
  selectors as paramsSelectors,
  actions as paramsActions,
} from '../../store/search/params';
import connectSearch from '../../hocs/connectSearch';
import RequireAdult from '../../components/RequireAdult/RequireAdult';

class Search extends React.Component {
  componentDidMount() {
    const {
      params,
      paramsState,
      searchParams,
      setParamsFromProps,
      loadPage,
      loadFirstPage,
    } = this.props;
    this.initialParamsState = R.merge(paramsState, params);
    setParamsFromProps(params);

    if (isNilOrEmpty(searchParams.query)) {
      loadPage(0);
    } else {
      // If query is set, it means the search form was submitted from another page
      // Later, if we serialize query value into URL, we should use a different logic.
      loadFirstPage();
    }
  }

  componentDidUpdate(prevProps) {
    const { params, indexName, searchParams, setParamsFromProps } = this.props;
    if (!R.equals(prevProps.params, params)) {
      setParamsFromProps(params);
    }

    const indexNameChanged = !R.equals(prevProps.indexName, indexName);
    const searchParamsChanged = !R.equals(prevProps.searchParams, searchParams);

    if (indexNameChanged || searchParamsChanged) {
      this.handleLoadFirstPage();
    }
  }

  handleLoadNextPage = () => {
    const { page, loadPage } = this.props;
    loadPage(page + 1);
  };

  handleLoadFirstPage() {
    const {
      loadFirstPage,
      mapParamsToUrl,
      paramsState,
      match,
      history,
    } = this.props;

    this.initialParamsState = null;
    loadFirstPage();

    if (mapParamsToUrl) {
      const nextUrl = mapParamsToUrl(paramsState);

      if (nextUrl !== match.url) {
        history.push(nextUrl);
      }
    }
  }

  render() {
    const { hits, currentCategory, setCategory, children } = this.props;

    return (
      <RequireAdult
        id="Search-confirmAdult"
        enabled={currentCategory && currentCategory.requireAdult}
        onReject={() => setCategory(null)}
      >
        {children({
          hits,
          loadNextPage: this.handleLoadNextPage,
        })}
      </RequireAdult>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  paramsState: paramsSelectors.paramsSelector,
  currentCategory: paramsSelectors.categoryObjectSelector(),
  searchParams: searchSelectors.searchParamsSelector,
  indexName: searchSelectors.indexNameSelector,
  hits: searchSelectors.hitsSelector,
  page: pageSelectors.pageSelector,
});

const mapDispatchToProps = {
  setParamsFromProps: paramsActions.setParamsFromProps,
  setCategory: paramsActions.setCategory,
  loadFirstPage: searchActions.loadFirstPage,
  loadPage: searchActions.loadPage,
};

export default R.compose(
  connectData({ categories: models.categories.all }),
  connectSearch(mapStateToProps, mapDispatchToProps),
  withRouter,
)(Search);
