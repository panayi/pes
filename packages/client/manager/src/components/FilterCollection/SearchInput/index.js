import React from 'react';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import pure from 'recompose/pure';
import SearchIcon from '@material-ui/icons/Search';
import { fade } from '@material-ui/core/styles/colorManipulator';
import withStyles from '@material-ui/core/styles/withStyles';

const styles = theme => ({
  wrapper: {
    flex: 1,
    padding: [0, theme.spacing.unit],
    position: 'relative',
    fontFamily: theme.typography.fontFamily,
    borderRadius: 2,
    background: fade(theme.palette.common.white, 0.15),
    '&:hover': {
      background: fade(theme.palette.common.white, 0.25),
    },
  },
  search: {
    width: theme.spacing.unit * 5,
    height: '100%',
    position: 'absolute',
    pointerEvents: 'none',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  input: {
    width: '100%',
    font: 'inherit',
    padding: `${theme.spacing.unit}px ${theme.spacing.unit}px ${
      theme.spacing.unit
    }px ${theme.spacing.unit * 5}px`,
    border: 0,
    display: 'block',
    verticalAlign: 'middle',
    whiteSpace: 'normal',
    background: 'none',
    margin: 0, // Reset for Safari
    color: 'inherit',
    '&:focus': {
      outline: 0,
    },
    '&::placeholder': {
      color: 'inherit',
      opacity: 0.5,
    },
  },
});

const SearchInput = ({ value, onChange, placeholder, classes }) => (
  <div className={classes.wrapper}>
    <div className={classes.search}>
      <SearchIcon />
    </div>
    <input
      className={classes.input}
      value={value}
      onChange={onChange}
      placeholder={placeholder}
    />
  </div>
);

SearchInput.propTypes = {
  classes: PropTypes.shape({}).isRequired,
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
};

SearchInput.defaultProps = {
  value: '',
  placeholder: '',
};

export default R.compose(
  withStyles(styles, {
    name: 'SearchInput',
  }),
  pure,
)(SearchInput);
