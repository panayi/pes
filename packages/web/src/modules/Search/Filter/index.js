import ByCategory from './Category/Category';
import ByLocation from './Location/Location';
import ByPrice from './Price/Price';
import SortBy from './SortBy/SortBy';

const Filter = {
  ByCategory,
  ByLocation,
  ByPrice,
  SortBy,
};

export default Filter;
