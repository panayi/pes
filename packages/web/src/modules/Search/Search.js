import React from 'react';
import * as R from 'ramda';
import { createStructuredSelector } from 'reselect';
import propsChanged from '@pesposa/core/src/utils/propsChanged';
import { actions as searchActions } from 'store/search';
import { selectors as hitsSelectors } from 'store/search/hits';
import { selectors as pageSelectors } from 'store/search/page';
import {
  selectors as categorySelectors,
  actions as categoryActions,
} from 'store/search/category';
import {
  selectors as ramParamsSelecotrs,
  actions as rawParamsActions,
} from 'store/search/rawParams';
import connectSearch from 'hocs/connectSearch';

class Search extends React.Component {
  componentWillMount() {
    const {
      category,
      currentCategory,
      setSelectedCategory,
      params,
      currentParams,
      setRawParams,
    } = this.props;
    if (category !== currentCategory) {
      setSelectedCategory(category);
    }

    if (!R.equals(params, currentParams)) {
      setRawParams(params);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.category !== this.props.category) {
      this.props.setSelectedCategory(nextProps.category);
    }

    if (propsChanged(['params'], this.props, nextProps)) {
      this.props.setRawParams(nextProps.params);
    }
  }

  render() {
    const { hits, page, loadPage, children } = this.props;

    return children({
      hits,
      page,
      loadPage,
    });
  }
}

const mapStateToProps = createStructuredSelector({
  hits: hitsSelectors.hitsSelector,
  page: pageSelectors.pageSelector,
  currentCategory: categorySelectors.categorySelector,
  currentParams: ramParamsSelecotrs.rawParamsSelector,
});

const mapDispatchToProps = {
  loadPage: searchActions.loadPage,
  setSelectedCategory: categoryActions.setCategory,
  setRawParams: rawParamsActions.setRawParams,
};

export default connectSearch(mapStateToProps, mapDispatchToProps)(Search);
