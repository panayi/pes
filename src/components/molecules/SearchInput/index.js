/* @flow */
import React from 'react';
import * as R from 'ramda';
import { withStyles } from 'material-ui';
import { fade } from 'material-ui/styles/colorManipulator';
import SearchIcon from 'material-ui-icons/Search';
import { connectSearchBox } from 'react-instantsearch/connectors';

type Props = {
  value: String,
  refine: Function,
}

// Based on:
// https://github.com/callemall/material-ui/blob/v1-beta/docs/src/modules/components/AppSearch.js

const styles = theme => ({
  wrapper: {
    fontFamily: theme.typography.fontFamily,
    position: 'relative',
    borderRadius: 2,
    background: fade(theme.palette.common.white, 0.15),
    '&:hover': {
      background: fade(theme.palette.common.white, 0.25),
    },
    '& $input': {
      transition: theme.transitions.create('width'),
      width: 200,
      '&:focus': {
        width: 250,
      },
    },
  },
  searchIcon: {
    width: theme.spacing.unit * 7,
    height: '100%',
    position: 'absolute',
    pointerEvents: 'none',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  input: {
    font: 'inherit',
    padding: `${theme.spacing.unit}px ${theme.spacing.unit}px ${theme.spacing.unit}px ${theme
      .spacing.unit * 7}px`,
    border: 0,
    display: 'block',
    verticalAlign: 'middle',
    whiteSpace: 'normal',
    background: 'none',
    margin: 0, // Reset for Safari
    color: 'inherit',
    width: '100%',
    '&:focus': {
      outline: 0,
    },
    '&::placeholder': {
      color: fade(theme.palette.common.white, 0.25),
    },
  },
});

export const SearchInput = ({ value, refine, classes }: Props) => (
  <div className={classes.wrapper}>
    <div className={classes.searchIcon}>
      <SearchIcon />
    </div>
    <input
      className={classes.input}
      value={value}
      onChange={e => refine(e.target.value)}
      placeholder="What are you looking for?"
    />
  </div>
);

export default R.compose(
  connectSearchBox,
  withStyles(styles),
)(SearchInput);
