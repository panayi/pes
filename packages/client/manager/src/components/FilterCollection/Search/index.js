import { isFunction } from 'ramda-adjunct';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose, withProps } from 'recompose';
import { actions, selectors } from 'store/filterCollection';
import SearchInput from '../SearchInput';
import withFilterCollectionId from '../withFilterCollectionId';

const mapStateToProps = createStructuredSelector({
  query: selectors.queryValueSelector,
});

const mapDispatchToProps = {
  setQuery: actions.setQuery,
};

export default compose(
  withFilterCollectionId,
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  withProps(({ id, query, setQuery, onChange }) => ({
    value: query,
    onChange: event => {
      if (isFunction(onChange)) {
        onChange(event);
      }

      return setQuery(id, event.target.value);
    },
  })),
)(SearchInput);
