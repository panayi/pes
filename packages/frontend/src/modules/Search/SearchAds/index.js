import Filter from './Filter';
import Query from './Query/Query';

const SearchAds = {
  Filter,
  Query,
};

export default SearchAds;
