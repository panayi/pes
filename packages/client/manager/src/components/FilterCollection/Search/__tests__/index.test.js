import React from 'react';
import { PropTypes } from 'prop-types';
import { withContext } from 'recompose';
import SearchInput from '../../SearchInput';
import Search from '../index';

describe('FilterCollection', () => {
  describe('Search', () => {
    const Wrapper = withContext(
      {
        filterCollectionId: PropTypes.string.isRequired,
      },
      ({ myId }) => ({
        filterCollectionId: myId,
      }),
    )(({ children }) => children);

    it('should render correctly', () => {
      const { component } = withMockStore(
        <Wrapper myId="myAwesomeId">
          <Search />
        </Wrapper>,
      );
      const wrapper = mount(component);
      expect(wrapper).toMatchSnapshot();
    });

    it('should get query from redux state', () => {
      const state = {
        filterCollection: {
          query: {
            myAwesomeId: 'myAwesomeQuery',
          },
        },
      };
      const { component } = withMockStore(
        <Wrapper myId="myAwesomeId">
          <Search />
        </Wrapper>,
        state,
      );
      const wrapper = mount(component);
      expect(wrapper.find(SearchInput).prop('value')).toEqual('myAwesomeQuery');
    });

    it('should handle onChange and call setQuery action', () => {
      const onChangeSpy = jest.fn();
      const { component, store } = withMockStore(
        <Wrapper myId="myAwesomeId">
          <Search onChange={onChangeSpy} />
        </Wrapper>,
      );
      const wrapper = mount(component);

      const event = { target: { name: 'query', value: 'myAwesomeQuery' } };
      wrapper.find('input').simulate('change', event);

      expect(onChangeSpy.mock.calls).toHaveLength(1);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
