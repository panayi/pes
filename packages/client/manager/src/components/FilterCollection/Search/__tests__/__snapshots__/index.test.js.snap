// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterCollection Search should handle onChange and call setQuery action 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "myAwesomeId",
      "value": "myAwesomeQuery",
    },
    "type": "filterCollection/setQuery",
  },
]
`;

exports[`FilterCollection Search should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <withContext(Component)
    myId="myAwesomeId"
  >
    <Component
      myId="myAwesomeId"
    >
      <withFilterCollectionId(Connect(withProps(WithStyles(pure(SearchInput)))))>
        <renameProp(Connect(withProps(WithStyles(pure(SearchInput)))))
          filterCollectionId="myAwesomeId"
        >
          <Connect(withProps(WithStyles(pure(SearchInput))))
            id="myAwesomeId"
          >
            <withProps(WithStyles(pure(SearchInput)))
              id="myAwesomeId"
              query=""
              setQuery={[Function]}
            >
              <WithStyles(pure(SearchInput))
                id="myAwesomeId"
                onChange={[Function]}
                query=""
                setQuery={[Function]}
                value=""
              >
                <pure(SearchInput)
                  classes={
                    Object {
                      "input": "SearchInput-input-3",
                      "search": "SearchInput-search-2",
                      "wrapper": "SearchInput-wrapper-1",
                    }
                  }
                  id="myAwesomeId"
                  onChange={[Function]}
                  query=""
                  setQuery={[Function]}
                  value=""
                >
                  <SearchInput
                    classes={
                      Object {
                        "input": "SearchInput-input-3",
                        "search": "SearchInput-search-2",
                        "wrapper": "SearchInput-wrapper-1",
                      }
                    }
                    id="myAwesomeId"
                    onChange={[Function]}
                    placeholder=""
                    query=""
                    setQuery={[Function]}
                    value=""
                  >
                    <div
                      className="SearchInput-wrapper-1"
                    >
                      <div
                        className="SearchInput-search-2"
                      >
                        <pure(Search)>
                          <Search>
                            <WithStyles(SvgIcon)>
                              <SvgIcon
                                classes={
                                  Object {
                                    "colorAction": "MuiSvgIcon-colorAction-7",
                                    "colorDisabled": "MuiSvgIcon-colorDisabled-8",
                                    "colorError": "MuiSvgIcon-colorError-9",
                                    "colorPrimary": "MuiSvgIcon-colorPrimary-5",
                                    "colorSecondary": "MuiSvgIcon-colorSecondary-6",
                                    "root": "MuiSvgIcon-root-4",
                                  }
                                }
                                color="inherit"
                                viewBox="0 0 24 24"
                              >
                                <svg
                                  aria-hidden="true"
                                  className="MuiSvgIcon-root-4"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                                  />
                                </svg>
                              </SvgIcon>
                            </WithStyles(SvgIcon)>
                          </Search>
                        </pure(Search)>
                      </div>
                      <input
                        className="SearchInput-input-3"
                        onChange={[Function]}
                        placeholder=""
                        value=""
                      />
                    </div>
                  </SearchInput>
                </pure(SearchInput)>
              </WithStyles(pure(SearchInput))>
            </withProps(WithStyles(pure(SearchInput)))>
          </Connect(withProps(WithStyles(pure(SearchInput))))>
        </renameProp(Connect(withProps(WithStyles(pure(SearchInput)))))>
      </withFilterCollectionId(Connect(withProps(WithStyles(pure(SearchInput)))))>
    </Component>
  </withContext(Component)>
</Provider>
`;
